### PR TITLE
Implementing r4/r5/r6 features

### PR DESCRIPTION
**Pull request by atangent**
# Description

Added all features of R4, R5, and R6. Fixed outdated frontend tests. 

Note: This PR is the result of many long pair-programming sessions with @SethDamiani and @atangent.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Pytest Tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: qa327/frontend.py [hunk 1]
```diff
@@ -186,9 +197,161 @@ def profile(user):
...
+    """
+    return True if ticket_name[0].isalnum() and ticket_name[-1].isalnum() and all(char.isalnum() or char.isspace() for char in ticket_name[1:-1]) and 6 <= len(ticket_name) <= 60 else False
+
+def is_quantity_of_tickets_valid(num_tickets):
+    """
+    Checks if the ticket quantity is valid according to the specifications.
+    :param num_tickets: the quantity of tickets to be tested
+    :returns: True if the ticket quantity satisfies all requirements
+    """
+    return True if num_tickets > 0 and num_tickets <= 100 else False
```
Comment (SethDamiani) [inline qa327/frontend.py]: These ternary operators can be removed to simplify return statements, e.g.
`return True if num_tickets > 0 and num_tickets <= 100 else False`
can be simplified to
`return num_tickets > 0 and num_tickets <= 100`.

This should probably be applied to all 6 input validation functions.

**Review by SethDamiani — CHANGES_REQUESTED** [review]
Once those unnecessary ternary operators are removed, it looks ready to merge.

File: qa327/frontend.py (hunk 1 above)
Comment (atangent) [reply qa327/frontend.py]: Very good call. Def makes the code cleaner

**Comment by atangent** [comment]
Looks good to me. I also made `is_ticket_name_valid` easier to read by separating the multiple conditions to `return True` by putting each `and` on separate lines and bracketing the diff conditions.

Copied from modified code in `frontend.py`:
```
return (ticket_name[0].isalnum()) and \
        (ticket_name[-1].isalnum()) and \
        (all(char.isalnum() or char.isspace() for char in ticket_name[1:-1])) and \
        (6 <= len(ticket_name) <= 60)
```

File: qa327/backend.py [hunk 2]
```diff
@@ -42,12 +43,70 @@ def register_user(email, name, password, password2):
...
+    :return: an error message if there is any, or None if register succeeds
+    """
+    ticket = get_ticket(ticket_id)
+    ticket.name = name
+    ticket.quantity = quantity
+    ticket.price = price
+    ticket.date = date
+    db.session.commit()
+
+def buy_ticket(ticket_id, buyer_id):
```
Comment (Chumingsun) [inline qa327/backend.py:78]: I would add docstrings for this function too, to match the `update_ticket` and `sell_ticket` functions that also have docstrings in the same file

**Review by Chumingsun — COMMENTED** [review]
After this minor fix, should be good to merge

**Review by SethDamiani — APPROVED** [review]

**Comment by atangent** [comment]
this is a BANGER. ready to merge. 😎 👍 💯

**Pull request merged**